### PR TITLE
Add webhook dispatcher deployment gear

### PR DESCRIPTION
**Pull request by AndreasBurger**

### CODE REVIEW

File: concourse/setup.py [hunk 1]
```diff
@@ -452,6 +452,38 @@ def destroy_concourse_landscape(config_name: str, release_name: str):
     namespace_helper.delete_namespace(namespace=release_name)
 
 
+def deploy_webhook_dispatcher_landscape(
+    webhook_dispatcher_cfg: WebhookDispatcher,
+    chart_dir: str,
+    deployment_name: str,
+):
+    chart_dir = os.path.abspath(chart_dir)
```
Comment (ccwienk) [inline concourse/setup.py]: does it need to exist?

File: concourse/setup.py [hunk 2]
```diff
@@ -552,6 +584,56 @@ def deploy_or_upgrade_concourse(
...
+
+    helm_env = os.environ.copy()
+    # set KUBECONFIG env-var in the copy to relative file path
+    helm_env['KUBECONFIG'] = KUBECONFIG_FILE_NAME
+
+    # create temp dir containing all previously referenced files
+    with tempfile.TemporaryDirectory() as temp_dir:
+        with open(os.path.join(temp_dir, WEBHOOK_DISPATCHER_HELM_VALUES_FILE_NAME), 'w') as f:
+            yaml.dump(create_webhook_dispatcher_helm_values(
+                webhook_dispatcher_cfg=webhook_dispatcher_cfg, cfg_factory=cfg_factory,), f
```
Comment (ccwienk) [inline concourse/setup.py]: place exactly one argument per line

File: concourse/setup.py [hunk 3]
```diff
@@ -552,6 +584,56 @@ def deploy_or_upgrade_concourse(
...
+        "--wait",
+        "--namespace", namespace,
+        # Use Helm's value-rendering mechanism to merge value-sources.
+        "--values", WEBHOOK_DISPATCHER_HELM_VALUES_FILE_NAME,
+        namespace, # release name is the same as namespace name
+        chart_dir,
+    ]
+
+    helm_env = os.environ.copy()
+    # set KUBECONFIG env-var in the copy to relative file path
```
Comment (ccwienk) [inline concourse/setup.py]: why use a relative path?

File: concourse/setup.py [hunk 4]
```diff
@@ -552,6 +584,56 @@ def deploy_or_upgrade_concourse(
...
+        chart_dir: str,
+        cfg_factory: ConfigFactory,
+        kubernetes_cfg: KubernetesConfig,
+        deployment_name: str='webhook-dispatcher',
+):
+    """Deploys (or upgrades) webhook dispatcher via Helm CLI"""
+    helm_executable = ensure_helm_setup()
+    chart_dir = util.existing_dir(chart_dir)
+
+    namespace = deployment_name
```
Comment (ccwienk) [inline concourse/setup.py]: not_none / empty?

File: cli/concourseutil.py [hunk 5]
```diff
@@ -316,3 +316,20 @@ def trigger_resource_check(
         pipeline_name=pipeline_name,
         resource_name=resource_name,
     )
+
+
+def deploy_or_upgrade_webhook_dispatcher(
+    webhook_dispatcher_name: CliHints.non_empty_string(help="webhook dispatcher config name"),
+    chart_dir: CliHints.existing_dir(help="directory of webhook dispatcher chart"),
```
Comment (ccwienk) [inline cli/concourseutil.py]: so this must be a directory in which the chart file exists? would it not be more intuitive to specify the helm-chart file instead?

File: charts/webhook-dispatcher/Chart.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v1
+name: webhook-dispatcher
+version: '0.1.0'
+kubeVersion: '>1.11.0'
+description: 'Todo'
```
Comment (ccwienk) [inline charts/webhook-dispatcher/Chart.yaml]: hmmm

File: charts/webhook-dispatcher/Chart.yaml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v1
```
Comment (ccwienk) [inline charts/webhook-dispatcher/Chart.yaml]: consistently quote strings (as is recommended practice for helm-charts afaik) - OR consistently _not_ quote string literals (both OK with me)

File: charts/webhook-dispatcher/templates/webhook-dispatcher-service.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: webhook-dispatcher-svc
+  labels:
+    app: webhook-dispatcher
+spec:
+  type: ClusterIP
+  ports:
+  - port: 80
```
Comment (ccwienk) [inline charts/webhook-dispatcher/templates/webhook-dispatcher-service.yaml]: ingress does tls-termination, right? so we have to listen on port 80, but the actual endpoint is accesible through port 443 / with TLS?

File: model/webhook_dispatcher.py [hunk 9]
```diff
@@ -21,5 +21,36 @@
...
+    def github_cfgs(self):
+        return self.raw.get('github_cfgs')
+
+    def tls_config(self):
+        return self.raw.get('tls_config')
+
+    def secrets_server(self):
+        return self.raw.get('secrets_server')
+
+    def kubeconfig(self):
```
Comment (ccwienk) [inline model/webhook_dispatcher.py]: If I properly recall, this attribute refers to a kubernetes-cfg name? if so, name it accordingly (kubernetes_cfg_name)

File: concourse/setup.py [hunk 10]
```diff
@@ -39,6 +39,9 @@
     ConcourseConfig,
     ConcourseApiVersion,
 )
+from model.webhook_dispatcher import (
+    WebhookDispatcher
```
Comment (ccwienk) [inline concourse/setup.py]: should be named WebhookDispatcherC(on)?f(i)?g for the sake of consistency

File: concourse/setup.py [hunk 11]
```diff
@@ -257,6 +260,49 @@ def create_instance_specific_helm_values(
...
 
 
+def create_webhook_dispatcher_helm_values(
+    webhook_dispatcher_cfg: WebhookDispatcher,
+    cfg_factory: ConfigFactory,
+):
+    not_none(webhook_dispatcher_cfg)
+    not_none(cfg_factory)
+
+    # create secrets server endpoint
```
Comment (ccwienk) [inline concourse/setup.py]: nit: calculate
also: please use SecretsServerConfig.endpoint_url() instead!

File: concourse/setup.py [hunk 12]
```diff
@@ -257,6 +260,49 @@ def create_instance_specific_helm_values(
     return instance_specific_values
 
 
+def create_webhook_dispatcher_helm_values(
+    webhook_dispatcher_cfg: WebhookDispatcher,
+    cfg_factory: ConfigFactory,
+):
+    not_none(webhook_dispatcher_cfg)
```
Comment (ccwienk) [inline concourse/setup.py]: use check_type

File: concourse/setup.py [hunk 13]
```diff
@@ -257,6 +260,49 @@ def create_instance_specific_helm_values(
...
+    not_none(cfg_factory)
```
Comment (ccwienk) [inline concourse/setup.py]: use check_type

Comment (ccwienk) [reply concourse/setup.py]: OR: annotate with @ensure_annotations

File: concourse/setup.py [hunk 14]
```diff
@@ -257,6 +260,49 @@ def create_instance_specific_helm_values(
...
+
+    # create secrets server endpoint
+    secrets_server_name = webhook_dispatcher_cfg.secrets_server()
+    secrets_server_cfg = cfg_factory.secrets_server(secrets_server_name)
+    secrets_server_endpoint = '/'.join([
+        secrets_server_cfg.endpoint_url(),
+        secrets_server_cfg.secrets().concourse_secret_name(),
+        secrets_server_cfg.secrets().concourse_attribute()])
+
+    concourse_cfg_names = webhook_dispatcher_cfg.concourse_cfgs()
```
Comment (ccwienk) [inline concourse/setup.py]: rename method to concourse_cfg_names. the method does reutrn cfg-names, so why name it differently?

File: concourse/setup.py [hunk 15]
```diff
@@ -257,6 +260,49 @@ def create_instance_specific_helm_values(
...
+    github_cfg_names = webhook_dispatcher_cfg.github_cfgs()
```
Comment (ccwienk) [inline concourse/setup.py]: ditto

File: concourse/setup.py [hunk 16]
```diff
@@ -257,6 +260,49 @@ def create_instance_specific_helm_values(
...
+    job_mapping_cfg_names = webhook_dispatcher_cfg.job_mappings()
```
Comment (ccwienk) [inline concourse/setup.py]: ditto

File: concourse/setup.py [hunk 17]
```diff
@@ -257,6 +260,49 @@ def create_instance_specific_helm_values(
...
+
+    env_vars = []
+    env_vars.append({
+        'name': 'SECRETS_SERVER_ENDPOINT', 'value': secrets_server_endpoint
```
Comment (ccwienk) [inline concourse/setup.py]: why? just pass the one relevant cfg_name (which would be the secrets-server cfg name)
also (but irrelevant): why not use a DICT?

File: concourse/setup.py [hunk 18]
```diff
@@ -552,6 +584,56 @@ def deploy_or_upgrade_concourse(
...
+        "--namespace", namespace,
+        # Use Helm's value-rendering mechanism to merge value-sources.
+        "--values", WEBHOOK_DISPATCHER_HELM_VALUES_FILE_NAME,
+        namespace, # release name is the same as namespace name
+        chart_dir,
+    ]
+
+    helm_env = os.environ.copy()
+    # set KUBECONFIG env-var in the copy to relative file path
+    helm_env['KUBECONFIG'] = KUBECONFIG_FILE_NAME
```
Comment (ccwienk) [inline concourse/setup.py]: we now have multiple places in this very module in which a helm subprocess is called with a patched environment to pass kubeconfig. do not continue code duplication. instead, please factor out a function to call a (helm/kubectl) subprocecess with patched env. pbly should be done as a ctx-mgr

File: concourse/setup.py (hunk 1 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: yes it must exists before

File: concourse/setup.py (hunk 2 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: done

File: concourse/setup.py (hunk 4 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: done

File: cli/concourseutil.py (hunk 5 above)
Comment (RaphaelVogel) [reply cli/concourseutil.py]: yes

File: charts/webhook-dispatcher/Chart.yaml (hunk 6 above)
Comment (RaphaelVogel) [reply charts/webhook-dispatcher/Chart.yaml]: done

File: charts/webhook-dispatcher/templates/webhook-dispatcher-service.yaml (hunk 8 above)
Comment (RaphaelVogel) [reply charts/webhook-dispatcher/templates/webhook-dispatcher-service.yaml]: 443 is terminated at ingress; forwards it to the service on port 80, which forwards it to 8080

File: model/webhook_dispatcher.py (hunk 9 above)
Comment (RaphaelVogel) [reply model/webhook_dispatcher.py]: Done

File: concourse/setup.py (hunk 10 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: Done

File: concourse/setup.py (hunk 11 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: Done

File: concourse/setup.py (hunk 12 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: Done

File: concourse/setup.py (hunk 13 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: done

File: concourse/setup.py (hunk 14 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: done

File: concourse/setup.py (hunk 15 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: done

File: concourse/setup.py (hunk 16 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: done

File: concourse/setup.py (hunk 17 above)
Comment (RaphaelVogel) [reply concourse/setup.py]: Done

File: concourse/setup.py (hunk 1 above)
Comment (ccwienk) [reply concourse/setup.py]: well.. how about util.existing_dir, then?

**Comment by ccwienk** [comment]
I cherry-picked (and partially ameneded) the commits from this PR into master. Thus closing.

**Pull request closed without merging**